File: app/src/main/java/org/docheinstein/minimote/packet/MinimotePacket.java
package org.docheinstein.minimote.packet;

import org.docheinstein.minimote.utils.ByteUtils;

public class MinimotePacket {

    public static final int HEADER_SIZE = 8;

    public static MinimotePacket fromData(byte[] data) {
        if (data == null)
            return null;

        if (data.length < HEADER_SIZE)
            return null;

        Long header = ByteUtils.toLong(data);

        if (header == null)
            return null;

        int packetLength = (int) ((header >>> 56) & 0xFF);
        MinimotePacketType packetType = MinimotePacketType.fromValue((int) ((header >>> 48) & 0xFF));
        long eventTime = header & 0xFFFFFFFFFFFFL;

        byte[] payload = null;

        if (data.length > HEADER_SIZE) {
            payload = new byte[packetLength - HEADER_SIZE];
            System.arraycopy(data, HEADER_SIZE, payload, 0, packetLength - HEADER_SIZE);
        }

        return new MinimotePacket(
                packetLength, packetType,
                eventTime, payload
        );
    }

    private int mPacketLength;
    private MinimotePacketType mPacketType;
    private long mEventTime;
    private byte[] mPayload;

    public MinimotePacket(int packetLength, MinimotePacketType packetType,
                          long eventTime, byte[] payload) {
        mPacketLength = packetLength;
        mPacketType = packetType;
        mEventTime = eventTime;
        mPayload = payload;
    }

    public MinimotePacket(int packetLength, MinimotePacketType packetType,
                          byte[] payload) {
        this(packetLength, packetType, System.currentTimeMillis(), payload);
    }

    public MinimotePacket(MinimotePacketType packetType) {
        this(HEADER_SIZE, packetType, System.currentTimeMillis(), null);
    }

    public MinimotePacket(MinimotePacketType packetType, long eventTime) {
        this(HEADER_SIZE, packetType, eventTime, null);
    }

    public int getPacketLength() { return mPacketLength; }
    public MinimotePacketType getEventType() { return mPacketType; }
    public long getEventTime() { return mEventTime; }
    public byte[] getPayload() { return mPayload; }
    public int getPayloadLength() { return mPayload != null ? mPayload.length : 0; }

    public byte[] toData() {
        long header =
            ((mPacketLength & 0xFFL) << 56) |
            ((mPacketType.value & 0xFFL) << 48) |
            mEventTime & 0xFFFFFFFFFFFFL;

        byte[] data = new byte[HEADER_SIZE + getPayloadLength()];
        ByteUtils.put64(header, data, 0);

        if (getPayloadLength() > 0)
            System.arraycopy(mPayload, 0, data, HEADER_SIZE, mPayload.length);

        return data;
    }

    @Override
    public String toString() {
        return  "Packet length: " + mPacketLength + " bytes" + "\n" +
                "Packet type: " + mPacketType + "\n" +
                "Event time: " + mEventTime + "\n" +
                "Payload: " + ByteUtils.toBinaryString(mPayload);
    }
}

File: app/src/main/java/org/docheinstein/minimotek/minimote/MinimotePacket.kt
package org.docheinstein.minimotek.minimote

import io.ktor.util.date.*
import org.docheinstein.minimotek.extensions.get8
import org.docheinstein.minimotek.extensions.set8
import org.docheinstein.minimotek.extensions.toBinaryString
import java.lang.Exception

/*
    | --- packet length --- | --- packet type --- | --- event time --- | ---- payload ---- |
    | ------- (1) --------- | ------ (1) -------- | ------- (6) -------| --- (variable) -- |
 */
class MinimotePacket(
    val packetLength: Int,
    val packetType: MinimotePacketType,
    val eventTime: Long,
    val payload: ByteArray
) {

    class InvalidPacketException(message: String) : Exception(message) {

    }

    companion object {
        const val HEADER_SIZE = 8

        fun fromData(data: ByteArray): MinimotePacket {
            if (data.size < HEADER_SIZE)
                throw InvalidPacketException("Invalid packet: must be at least 8 bytes, but it's ${data.size} bytes")

            val header: Long = data.get8()

            val packetLength = ((header ushr 56).toInt() and 0xFF)
            val packetTypeVal = ((header ushr 48).toInt() and 0xFF)
            val packetType = MinimotePacketType[packetTypeVal]
                ?: throw InvalidPacketException("Invalid packet: unknown packet type ${packetTypeVal.toString(16)}")

            val eventTime = header and 0xFFFFFFFFFFFFL

            val payload: ByteArray
            if (packetLength > HEADER_SIZE) {
                payload = ByteArray(packetLength - HEADER_SIZE)
                data.copyInto(payload, 0, HEADER_SIZE, packetLength)
            } else {
                payload = ByteArray(0)
            }

            return MinimotePacket(packetLength, packetType, eventTime, payload)
        }
    }

    constructor(packetLength: Int, packetType: MinimotePacketType, payload: ByteArray) : this(
        packetLength, packetType, getTimeMillis(), payload
    )

    constructor(packetType: MinimotePacketType) : this(
        HEADER_SIZE, packetType, getTimeMillis(), ByteArray(0)
    )


    fun toBytes(): ByteArray {
        val data = ByteArray(packetLength)

        val header: Long =
            ((packetLength and 0xFF).toLong() shl 56) or
            ((packetType.value and 0xFF).toLong() shl 48) or
            (eventTime and 0xFFFFFFFFFFFFL)

        data.set8(header)

        if (payload.isNotEmpty())
            payload.copyInto(data, HEADER_SIZE)

        return data
    }

    override fun toString(): String {
        return """
Packet length: $packetLength bytes
Packet type: $packetType
Event time: $eventTime
Payload: ${payload.toBinaryString(pretty = true)}
Dump: ${toBytes().toBinaryString(pretty = true)}
"""
    }
}